File: Java/com/leetcode/binaryTreeLevelOrderTraversalII/BinaryTreeLevelOrderTraversalII.java
package com.leetcode.binaryTreeLevelOrderTraversalII;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import com.leetcode.dataStructures.TreeNode;

public class BinaryTreeLevelOrderTraversalII {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        LinkedList<ArrayList<Integer>> traversal = new LinkedList<ArrayList<Integer>>();
        
        if (root == null) {
            return (List)traversal;
        }
        
        LinkedList<TreeNode> queue = new LinkedList<TreeNode>();
        TreeNode dummy = new TreeNode(-1);
        ArrayList<Integer> level = new ArrayList<Integer>();
        
        queue.add(root);
        queue.add(dummy);
        int length = 2;
        while (length > 0) {
            TreeNode n = queue.pop();
            --length;
            
            if (n == dummy) {
                traversal.push(level);
                if (length > 0) {
                    level = new ArrayList<Integer>();
                    queue.add(dummy);
                    ++length;
                }
            } else {
                level.add(n.val);
                if (n.left != null) {
                    queue.add(n.left);
                    ++length;
                }
                if (n.right != null) {
                    queue.add(n.right);
                    ++length;
                }
            }
        }
        
        return (List)traversal;
    }
}

File: Python/binary_tree_level_order_traversal_ii/binary_tree_level_order_traversal_ii.py
class Solution(object):
	def levelOrderBottom(self, root):
		if root is None:
			return []
		res = []
		lvl = 0
		curr_l = lvl
		q = [(root, lvl)]
		while q:
			(rt, l) = q.pop(0)
			if curr_l == l and res != []:
				res[-1].append(rt.val)
			else:
				res.append([rt.val])
				curr_l = l
			if rt.left:
				q.append((rt.left, l+1))
			if rt.right:
				q.append((rt.right, l+1))
		return list(reversed(res))
# TODO: Queue.Queue
